File: src/main/java/org/jukeboxmc/block/BlockJigsaw.java
package org.jukeboxmc.block;

import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.item.ItemJigsaw;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockJigsaw extends Block {

    public BlockJigsaw() {
        super( "minecraft:jigsaw" );
    }

    @Override
    public ItemJigsaw toItem() {
        return new ItemJigsaw();
    }

    @Override
    public BlockType getType() {
        return BlockType.JIGSAW;
    }

    public void setRotation( int value ) { //0-3
        this.setState( "rotation", value );
    }

    public int getRotation() {
        return this.stateExists( "rotation" ) ? this.getIntState( "rotation" ) : 0;
    }

    public void setBlockFace( BlockFace blockFace ) {
        this.setState( "facing_direction", blockFace.ordinal() );
    }

    public BlockFace getBlockFace() {
        return this.stateExists( "facing_direction" ) ? BlockFace.values()[this.getIntState( "facing_direction" )] : BlockFace.NORTH;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockJigsaw.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.Jigsaw
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.server.block.JukeboxBlock

class BlockJigsaw(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates), Jigsaw {

   override fun getFacingDirection(): BlockFace {
       return BlockFace.entries[this.getIntState("facing_direction")]
   }

   override fun setFacingDirection(value: BlockFace): Jigsaw {
       return this.setState("facing_direction", value.ordinal)
   }

   override fun getRotation(): Int {
       return this.getIntState("rotation")
   }

   override fun setRotation(value: Int): Jigsaw {
       return this.setState("rotation", value)
   }
}